Keep host port socket open for kubenet

CaoShuFeng opened the pull request:
fixes #37087


**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
NONE
```

When cni is set to kubenet, kubelet should hold the host port socket,
so that other application in this node could not listen/bind this port
any more. However, the sockets are closed accidentally, because
kubelet forget to reconcile the protocol format before comparing.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37088)
<!-- Reviewable:end -->
@kubernetes/sig-network

dims commented:
@k8s-bot ok to test

CaoShuFeng commented:
Sorry, I do a wrong rebase action.
So, many unrelated changes show up in this pr.
I will try to fix it.

CaoShuFeng commented:
@k8s-bot verify test this

CaoShuFeng commented:
@k8s-bot verify test this

feiskyer commented:
@k8s-bot verify test this

feiskyer commented on pkg/kubelet/network/hostport/hostport_test.go:
```diff
@@ -185,6 +186,10 @@ func TestOpenPodHostports(t *testing.T) {
 
+	// Already running pod's host port
+	hp := hostport{tests[1].pod.Spec.Containers[0].Ports[0].HostPort, strings.ToLower(string(tests[1].pod.Spec.Containers[0].Ports[0].Protocol))}
+	h.hostPortMap[hp] = &fakeSocket{tests[1].pod.Spec.Containers[0].Ports[0].HostPort, strings.ToLower(string(tests[1].pod.Spec.Containers[0].Ports[0].Protocol)), false}
```
nit: lines are too long, consider to multiple lines.

CaoShuFeng replied:
Done.

caseydavenport commented on pkg/kubelet/network/hostport/hostport.go:
```diff
@@ -97,7 +97,8 @@ func (h *handler) openHostports(pod *api.Pod) error {
 			hp := hostport{
-				port:     port.HostPort,
+				port: port.HostPort,
+				// When cleaning it up, we also check the protocal in lower format.
```
Can you please undo this change?

CaoShuFeng replied:
Done.

freehan commented:
Thanks! LGTM

freehan commented:
Applying 1.5 milestone since this is a bug fix.

saad-ali commented:
> Applying 1.5 milestone since this is a bug fix.

Ack.

feiskyer commented:
@k8s-bot gce etcd3 e2e test this

CaoShuFeng commented:
Hi, the ci test says verify-gofmt.sh failed, but I have passed it in my local environment, can someone help me?
>I1119 13:15:47.721] !!! 'gofmt -s -w' needs to be run on the following files: 
>I1119 13:15:47.721] ./pkg/kubelet/network/hostport/hostport.go
>I1119 13:15:47.722] ./pkg/kubelet/network/hostport/hostport_test.go
>I1119 13:15:47.723] FAILED   hack/make-rules/../../hack/verify-gofmt.sh	9s

feiskyer commented:
@CaoShuFeng Maybe try rebase and gofmt again?

CaoShuFeng commented:
@feiskyer rebased and run "bash hack/verify-gofmt.sh" also ok.

freehan commented:
@CaoShuFeng did you try a different go/gofmt version?

CaoShuFeng commented:
I rebased it.
And run "make verify" "make test" successfully in my environment.
Hope this time it works.

CaoShuFeng committed:
Keep host port socket open for kubenet

When cni is set to kubenet, kubelet should hold the host port socket,
so that other application in this node could not listen/bind this port
any more. However, the sockets are closed accidentally, because
kubelet forget to reconcile the protocol format before comparing.

dims commented:
@k8s-bot test this

dims commented:
@k8s-bot test this

dims commented:
Dear @kubernetes/test-infra-maintainers Where can i look up why 10 of the jobs ended with "Error in queue"?

krzyzacy commented:
the error message come from here: https://github.com/kubernetes/test-infra/blob/58bb3baabf5a67f0c86a8f73516d1d8c5d625303/prow/cmd/line/main.go#L285

maybe a jenkins request error, and seems it's after https://github.com/kubernetes/test-infra/pull/1215 is merged. @spxtr probably can explain better.

dims commented:
@k8s-bot node e2e test this
@k8s-bot cvm gce e2e test this
@k8s-bot unit test this
@k8s-bot verify test this
@k8s-bot node e2e test this
@k8s-bot kubemark e2e test this
@k8s-bot gce etcd3 e2e test this

dims commented:
Requesting the minimum jobs needed for this to be picked up by SQ

spxtr commented:
@dims feel free to open an issue in test-infra for this sort of thing in the future. We're looking into the "Error in queue" issue and hopefully it's fixed by now.

dims commented:
Ack @spxtr will do!

dims commented:
@k8s-bot test this [Multiple CI jobs failed]

rmmh commented:
@k8s-bot test this issue kubernetes/test-infra#1253 (also, deleting spurious failure comments)

dims commented:
@k8s-bot gce etcd3 e2e test this
@k8s-bot gci gke e2e test this

saad-ali commented:
Does this need to be cherry-picked to 1.5? If so make sure to add `cherrypick-candidate` label

saad-ali commented:
> Does this need to be cherry-picked to 1.5? If so make sure to add cherrypick-candidate label

Ping. CC @freehan
